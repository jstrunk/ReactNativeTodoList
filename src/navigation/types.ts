import { StackNavigationProp } from "@react-navigation/stack";
import { RouteProp } from "@react-navigation/native";

export type RootStackParamList = {
  TodoSelect: undefined;
  TodoList: {id: string};
};

export type TodoSelectScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'TodoSelect'
>;

export type TodoListScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'TodoList'
>;

export type TodoListScreenRouteProp = RouteProp<
  RootStackParamList,
  'TodoList'
>;

