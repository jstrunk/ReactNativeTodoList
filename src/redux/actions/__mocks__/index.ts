export const addItem = jest.fn();
export const toggleItem = jest.fn();
export const deleteItem = jest.fn(); 
export const deleteCompleted = jest.fn(); 
export const reorderList = jest.fn(); 
export const registerList = jest.fn(); 
export const setActiveList = jest.fn(); 
export const createList = jest.fn(); 
export const loadList = jest.fn(); 