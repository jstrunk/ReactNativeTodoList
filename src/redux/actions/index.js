import { ADD_ITEM, DELETE_ITEM, TOGGLE_ITEM, DELETE_COMPLETED } from '../constants';

export function addItem(text) {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index) {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index) {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted() {
  return { type: DELETE_COMPLETED };
}
