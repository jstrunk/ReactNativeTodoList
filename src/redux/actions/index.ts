import { ADD_ITEM,
  DELETE_ITEM,
  TOGGLE_ITEM,
  DELETE_COMPLETED,
  TodoActionType, } from '../types';

export function addItem(text: string): TodoActionType {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index: number): TodoActionType {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index: number): TodoActionType {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted(): TodoActionType {
  return { type: DELETE_COMPLETED };
}
