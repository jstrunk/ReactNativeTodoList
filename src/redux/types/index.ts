export interface Todo {
  key: string;
  done: boolean;
}

export interface ITodoList {
  byId: Record<string,Todo>;
  allItems: Array<string>;
}

export interface ITodoState {
  todoList: ITodoList;
}

export const defaultState: ITodoList = {
  byId: {},
  allItems: [],
};

export const ADD_ITEM = "ADD_ITEM";
export const TOGGLE_ITEM = "TOGGLE_ITEM";
export const DELETE_ITEM = "DELETE_ITEM";
export const DELETE_COMPLETED = "DELETE_COMPLETED";

interface AddItemAction {
  type: typeof ADD_ITEM;
  text: string;
  id: string;
}

interface ToggleItemAction {
  type: typeof TOGGLE_ITEM;
  id: string;
}

interface DeleteItemAction {
  type: typeof DELETE_ITEM;
  id: string;
}

interface DeleteCompletedAction {
  type: typeof DELETE_COMPLETED;
  ids: Array<string>;
}

export type TodoActionType = AddItemAction | ToggleItemAction |
  DeleteItemAction | DeleteCompletedAction;
