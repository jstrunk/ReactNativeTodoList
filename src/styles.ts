import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
  },
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 22
  },
  modalView: {
    margin: 20,
    backgroundColor: 'lightgrey',
    borderRadius: 20,
    padding: 35,
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
    width: '75%',
  },
  footer: {
    backgroundColor: 'white',
    padding: 15,
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
  },
  remove: {
    textAlign: 'center',
    color: 'red',
  },
  addItem: {
    padding: 15,
    textAlign: 'left',
    color: 'grey',
    borderBottomColor: 'lightgrey',
    borderBottomWidth: 1,
    width: '100%',
  },
  list: {
    marginBottom: 40,
  },
  item: {
    textAlign: 'left',
    color: 'black',
  },
  itemdone: {
    textAlign: 'left',
    color: 'black',
    textDecorationLine: 'line-through',
    textDecorationStyle: 'solid',
  },
  checkBox: {
    color: '#bfbfbf',
  },
  header: {
    backgroundColor: 'skyblue',
  },
  title: {
    textAlign: 'center',
    color: 'white',
  },
  backbutton: {
    textAlign: 'left',
    color: 'white',
  },
  createListButtonView: {
    position: 'absolute',
    right: 50,
    bottom: 80,
  },
  createListTextInput: {
    padding: 15,
    textAlign: 'left',
    borderBottomWidth: 1,
    width: '100%',
  },
});

export default styles;