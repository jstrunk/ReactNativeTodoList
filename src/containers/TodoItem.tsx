/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React from 'react';
import { Pressable, StyleSheet, Text } from 'react-native';
import { connect, ConnectedProps } from 'react-redux';
import { toggleItem, deleteItem } from '../redux/actions';
import { ITodoState } from '../redux/types';
import { ListItem } from 'react-native-elements';

const styles = StyleSheet.create({
  item: {
    textAlign: 'left',
    color: 'black',
  },
  remove: {
    textAlign: 'center',
    color: 'red',
  },
  checkBox: {
    color: '#bfbfbf',
  },
});

interface OwnProps {
  id: string;
  styles?: typeof styles;
}

const mapStateToProps = (state: ITodoState) => ({
  items: state.todoItems,
})

const mapDispatch = {
  toggleItem: (id: string) => toggleItem(id),
  deleteItem: (id: string) => deleteItem(id),
};

const connector = connect(
  mapStateToProps,
  mapDispatch
);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = OwnProps & PropsFromRedux;

const TodoItem = (props: Props) => {
  const item = props.items[props.id];
  const style = props.styles ? props.styles : styles;
  return (
    <ListItem
      title={item.key}
      titleStyle={style.item}
      checkBox={{
        right: true,
        checked: item.done,
        onIconPress: () => props.toggleItem(props.id),
      }}
      rightElement={<Pressable onPress={() => props.deleteItem(props.id)}>
          <Text style={style.remove}>X</Text>
        </Pressable>
        }
    />
  );
}

export default connector(TodoItem);
