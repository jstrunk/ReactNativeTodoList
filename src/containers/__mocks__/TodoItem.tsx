/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React from 'react';
import { ListItem } from 'react-native-elements';
import styles from '../../styles';

interface OwnProps {
  id: string;
}

type Props = OwnProps;

const TodoItem = (props: Props): JSX.Element => {
  return (
    <ListItem
      title={props.id}
      titleStyle={styles.item}
    />
  );
}

export default TodoItem;
