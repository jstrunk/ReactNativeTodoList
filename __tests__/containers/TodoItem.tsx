import React from 'react';
import TodoItem from '../../src/containers/TodoItem';
import * as types from '../../src/redux/types';
import * as actions from '../../src/redux/actions';
import { Provider } from 'react-redux';
import { render, fireEvent } from '@testing-library/react-native'
import styles from '../../src/styles';
import configureMockStore, { MockStore } from 'redux-mock-store';
import thunk, { ThunkDispatch } from 'redux-thunk';
import { oneListOneItemDone } from '../../test/testStates';

jest.mock('../../src/redux/actions');

type DispatchExts = ThunkDispatch<types.ITodoState, void, types.CombinedActionType>;
const middlewares = [thunk];
const mockStore = configureMockStore<types.ITodoState,DispatchExts>(middlewares);


describe('TodoItem container', () => {
  let store: MockStore;

  beforeEach(() => {
    store = mockStore(oneListOneItemDone);

    store.dispatch = jest.fn();
  });

  it('renders a TodoItem', () => {
    const { toJSON } = render(
      <Provider store={store}>
        <TodoItem id='abc1' styles={styles} />
      </Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('has items', () => {
    const { getAllByText } = render(
      <Provider store={store}>
        <TodoItem id='def2' />
      </Provider >
    );

    const fooElements = getAllByText('foo');
    expect(fooElements).toHaveLength(1);
  });

  it('should dispatch a toggle item action', () => {
    const { getAllByText } = render(
      <Provider store={store}>
        <TodoItem id='abc1' />
      </Provider>
    );

    const checkedElements = getAllByText('');
    expect(checkedElements).toHaveLength(1);

    checkedElements.forEach((e) => fireEvent.press(e));

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      actions.toggleItem('abc1')
    );
  });

  it('should dispatch a delete item action', () => {
    const { getAllByText } = render(
      <Provider store={store}>
        <TodoItem id='abc1' />
      </Provider>
    );

    const deleteElements = getAllByText('X');
    expect(deleteElements).toHaveLength(1);

    deleteElements.forEach((e) => fireEvent.press(e));

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      actions.deleteItem('abc1')
    );
    expect(actions.deleteItem).toHaveBeenCalledWith('abc1');
  });
});
