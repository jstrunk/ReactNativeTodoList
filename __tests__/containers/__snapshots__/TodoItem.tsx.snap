// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem container renders a TodoItem 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderColor": "#bcbbc1",
        "flexDirection": "row",
        "padding": 14,
      }
    }
    testID="padView"
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "transparent",
            "color": "black",
            "fontSize": 17,
            "textAlign": "left",
            "textDecorationLine": "line-through",
            "textDecorationStyle": "solid",
          }
        }
        testID="listItemTitle"
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
      >
        Run the tests
      </Text>
    </View>
    <View
      style={
        Object {
          "paddingLeft": 16,
        }
      }
    />
    <View
      accessibilityRole="checkbox"
      accessibilityState={
        Object {
          "checked": true,
        }
      }
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "margin": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "opacity": 1,
          "padding": 0,
        }
      }
      testID="checkbox"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <Text
          allowFontScaling={false}
          onPress={[Function]}
          style={
            Array [
              Object {
                "color": "#2089dc",
                "fontSize": 24,
              },
              Object {
                "minWidth": 24,
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ÔÅÜ
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "paddingLeft": 16,
        }
      }
    />
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          Object {
            "color": "red",
            "textAlign": "center",
          }
        }
      >
        X
      </Text>
    </View>
  </View>
</View>
`;
